File: epee/epee.go
// Package epee implements a minimal encoder/decoder for epee's "portable
// storage" binary format — the serialization used by Zano's daemon .bin RPC
// endpoints (and by wallet key blobs). Only the subset needed by this library
// is implemented: ordered sections, unsigned integers, booleans, strings/blobs,
// nested objects, and arrays of integers or objects.
//
// Reference: zano/contrib/epee/include/storages/portable_storage*.h
package epee

import (
	"encoding/binary"
	"fmt"
)

const (
	signatureA = 0x01011101
	signatureB = 0x01020101
	formatVer  = 1

	typeInt64  = 1
	typeInt32  = 2
	typeInt16  = 3
	typeInt8   = 4
	typeUint64 = 5
	typeUint32 = 6
	typeUint16 = 7
	typeUint8  = 8
	typeDouble = 9
	typeString = 10
	typeBool   = 11
	typeObject = 12
	typeArray  = 13

	flagArray = 0x80
)

// Section is an ordered set of named entries (an epee "section"/object). Order
// is preserved for deterministic encoding; lookup is by name.
type Section struct {
	keys   []string
	values map[string]any
}

// NewSection returns an empty Section.
func NewSection() *Section {
	return &Section{values: map[string]any{}}
}

// Set adds or replaces an entry. Supported value types: uint64, bool, string,
// []byte, *Section, []uint64, []*Section.
func (s *Section) Set(name string, v any) *Section {
	if _, ok := s.values[name]; !ok {
		s.keys = append(s.keys, name)
	}
	s.values[name] = v
	return s
}

// Get returns the raw value for name (as produced by Unmarshal or Set).
func (s *Section) Get(name string) (any, bool) {
	v, ok := s.values[name]
	return v, ok
}

// Uint64 returns a uint64-typed entry.
func (s *Section) Uint64(name string) (uint64, bool) {
	v, ok := s.values[name]
	if !ok {
		return 0, false
	}
	u, ok := v.(uint64)
	return u, ok
}

// Bytes returns a string/blob entry as bytes.
func (s *Section) Bytes(name string) ([]byte, bool) {
	v, ok := s.values[name]
	if !ok {
		return nil, false
	}
	if b, ok := v.([]byte); ok {
		return b, true
	}
	if str, ok := v.(string); ok {
		return []byte(str), true
	}
	return nil, false
}

// Sections returns an array-of-object entry.
func (s *Section) Sections(name string) ([]*Section, bool) {
	v, ok := s.values[name]
	if !ok {
		return nil, false
	}
	a, ok := v.([]*Section)
	return a, ok
}

// Marshal encodes a root section to the portable-storage binary format.
func Marshal(root *Section) []byte {
	var b []byte
	b = binary.LittleEndian.AppendUint32(b, signatureA)
	b = binary.LittleEndian.AppendUint32(b, signatureB)
	b = append(b, formatVer)
	b = appendSection(b, root)
	return b
}

func appendVarint(b []byte, v uint64) []byte {
	switch {
	case v <= 63:
		return append(b, byte(v<<2))
	case v <= 16383:
		return binary.LittleEndian.AppendUint16(b, uint16(v<<2)|1)
	case v <= 1073741823:
		return binary.LittleEndian.AppendUint32(b, uint32(v<<2)|2)
	default:
		return binary.LittleEndian.AppendUint64(b, (v<<2)|3)
	}
}

func appendString(b, s []byte) []byte {
	b = appendVarint(b, uint64(len(s)))
	return append(b, s...)
}

func appendSection(b []byte, s *Section) []byte {
	b = appendVarint(b, uint64(len(s.keys)))
	for _, k := range s.keys {
		b = append(b, byte(len(k)))
		b = append(b, k...)
		b = appendEntry(b, s.values[k])
	}
	return b
}

func appendEntry(b []byte, v any) []byte {
	switch x := v.(type) {
	case uint64:
		b = append(b, typeUint64)
		return binary.LittleEndian.AppendUint64(b, x)
	case bool:
		b = append(b, typeBool)
		if x {
			return append(b, 1)
		}
		return append(b, 0)
	case string:
		b = append(b, typeString)
		return appendString(b, []byte(x))
	case []byte:
		b = append(b, typeString)
		return appendString(b, x)
	case *Section:
		b = append(b, typeObject)
		return appendSection(b, x)
	case []uint64:
		b = append(b, typeUint64|flagArray)
		b = appendVarint(b, uint64(len(x)))
		for _, e := range x {
			b = binary.LittleEndian.AppendUint64(b, e)
		}
		return b
	case []*Section:
		b = append(b, typeObject|flagArray)
		b = appendVarint(b, uint64(len(x)))
		for _, e := range x {
			b = appendSection(b, e)
		}
		return b
	default:
		panic(fmt.Sprintf("epee: unsupported value type %T", v))
	}
}

// reader is a cursor over the encoded body (after the 9-byte header).
type reader struct {
	b []byte
	p int
}

func (r *reader) need(n int) error {
	if r.p+n > len(r.b) {
		return fmt.Errorf("epee: unexpected end of buffer (need %d at %d of %d)", n, r.p, len(r.b))
	}
	return nil
}

func (r *reader) byte() (byte, error) {
	if err := r.need(1); err != nil {
		return 0, err
	}
	v := r.b[r.p]
	r.p++
	return v, nil
}

func (r *reader) take(n int) ([]byte, error) {
	if err := r.need(n); err != nil {
		return nil, err
	}
	v := r.b[r.p : r.p+n]
	r.p += n
	return v, nil
}

func (r *reader) varint() (uint64, error) {
	first, err := r.byte()
	if err != nil {
		return 0, err
	}
	mask := first & 0x03
	r.p-- // re-read including the first byte
	switch mask {
	case 0:
		v, err := r.take(1)
		if err != nil {
			return 0, err
		}
		return uint64(v[0]) >> 2, nil
	case 1:
		v, err := r.take(2)
		if err != nil {
			return 0, err
		}
		return uint64(binary.LittleEndian.Uint16(v)) >> 2, nil
	case 2:
		v, err := r.take(4)
		if err != nil {
			return 0, err
		}
		return uint64(binary.LittleEndian.Uint32(v)) >> 2, nil
	default:
		v, err := r.take(8)
		if err != nil {
			return 0, err
		}
		return binary.LittleEndian.Uint64(v) >> 2, nil
	}
}

// Unmarshal decodes a portable-storage binary buffer into its root Section.
func Unmarshal(data []byte) (*Section, error) {
	if len(data) < 9 {
		return nil, fmt.Errorf("epee: too short (%d < 9)", len(data))
	}
	if binary.LittleEndian.Uint32(data[0:4]) != signatureA || binary.LittleEndian.Uint32(data[4:8]) != signatureB {
		return nil, fmt.Errorf("epee: bad signature")
	}
	if data[8] != formatVer {
		return nil, fmt.Errorf("epee: unknown format version %d", data[8])
	}
	r := &reader{b: data[9:]}
	return r.readSection()
}

func (r *reader) readSection() (*Section, error) {
	count, err := r.varint()
	if err != nil {
		return nil, err
	}
	s := NewSection()
	for i := uint64(0); i < count; i++ {
		nl, err := r.byte()
		if err != nil {
			return nil, err
		}
		name, err := r.take(int(nl))
		if err != nil {
			return nil, err
		}
		v, err := r.readEntry()
		if err != nil {
			return nil, err
		}
		s.Set(string(name), v)
	}
	return s, nil
}

func (r *reader) readEntry() (any, error) {
	t, err := r.byte()
	if err != nil {
		return nil, err
	}
	if t&flagArray != 0 {
		return r.readArray(t &^ flagArray)
	}
	return r.readValue(t)
}

func (r *reader) readArray(base byte) (any, error) {
	count, err := r.varint()
	if err != nil {
		return nil, err
	}
	if base == typeObject {
		out := make([]*Section, 0, count)
		for i := uint64(0); i < count; i++ {
			sec, err := r.readSection()
			if err != nil {
				return nil, err
			}
			out = append(out, sec)
		}
		return out, nil
	}
	out := make([]any, 0, count)
	for i := uint64(0); i < count; i++ {
		v, err := r.readValue(base)
		if err != nil {
			return nil, err
		}
		out = append(out, v)
	}
	return out, nil
}

func (r *reader) readValue(t byte) (any, error) {
	switch t {
	case typeUint64, typeInt64:
		v, err := r.take(8)
		if err != nil {
			return nil, err
		}
		return binary.LittleEndian.Uint64(v), nil
	case typeUint32, typeInt32:
		v, err := r.take(4)
		if err != nil {
			return nil, err
		}
		return uint64(binary.LittleEndian.Uint32(v)), nil
	case typeUint16, typeInt16:
		v, err := r.take(2)
		if err != nil {
			return nil, err
		}
		return uint64(binary.LittleEndian.Uint16(v)), nil
	case typeUint8, typeInt8:
		v, err := r.byte()
		if err != nil {
			return nil, err
		}
		return uint64(v), nil
	case typeBool:
		v, err := r.byte()
		if err != nil {
			return nil, err
		}
		return v != 0, nil
	case typeDouble:
		v, err := r.take(8)
		if err != nil {
			return nil, err
		}
		return binary.LittleEndian.Uint64(v), nil // raw bits; not needed by us
	case typeString:
		n, err := r.varint()
		if err != nil {
			return nil, err
		}
		v, err := r.take(int(n))
		if err != nil {
			return nil, err
		}
		// return a copy as []byte so callers get exact bytes (may be binary)
		out := make([]byte, len(v))
		copy(out, v)
		return out, nil
	case typeObject:
		return r.readSection()
	default:
		return nil, fmt.Errorf("epee: unsupported value type %d", t)
	}
}

File: src/epee.rs
//! A minimal encoder/decoder for epee's "portable storage" binary format — the
//! serialization used by Zano's daemon `.bin` RPC endpoints.
//!
//! Only the subset this library needs is implemented: ordered sections,
//! unsigned integers, booleans, strings/blobs, nested objects, and arrays of
//! integers or objects.
//!
//! Reference: `zano/contrib/epee/include/storages/portable_storage*.h`.

use crate::error::{Error, Result};

const SIGNATURE_A: u32 = 0x0101_1101;
const SIGNATURE_B: u32 = 0x0102_0101;
const FORMAT_VER: u8 = 1;

const TYPE_INT64: u8 = 1;
const TYPE_INT32: u8 = 2;
const TYPE_INT16: u8 = 3;
const TYPE_INT8: u8 = 4;
const TYPE_UINT64: u8 = 5;
const TYPE_UINT32: u8 = 6;
const TYPE_UINT16: u8 = 7;
const TYPE_UINT8: u8 = 8;
const TYPE_DOUBLE: u8 = 9;
const TYPE_STRING: u8 = 10;
const TYPE_BOOL: u8 = 11;
const TYPE_OBJECT: u8 = 12;

const FLAG_ARRAY: u8 = 0x80;

/// A value in an epee section.
#[derive(Clone, Debug, PartialEq)]
pub enum Value {
    /// An unsigned integer (every integer width decodes into this).
    Uint(u64),
    /// A boolean.
    Bool(bool),
    /// A string or binary blob.
    Bytes(Vec<u8>),
    /// A nested section.
    Section(Section),
    /// An array of integers.
    UintArray(Vec<u64>),
    /// An array of sections.
    SectionArray(Vec<Section>),
    /// An array of any other value type.
    Array(Vec<Value>),
}

/// An ordered set of named entries (an epee "section"/object).
///
/// Insertion order is preserved so encoding is deterministic; lookup is by name.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct Section {
    entries: Vec<(String, Value)>,
}

impl Section {
    /// An empty section.
    pub fn new() -> Section {
        Section::default()
    }

    /// Adds or replaces an entry, returning `self` for chaining.
    pub fn set(&mut self, name: &str, v: impl Into<Value>) -> &mut Section {
        let v = v.into();
        match self.entries.iter_mut().find(|(k, _)| k == name) {
            Some(slot) => slot.1 = v,
            None => self.entries.push((name.to_string(), v)),
        }
        self
    }

    /// The raw value for `name`.
    pub fn get(&self, name: &str) -> Option<&Value> {
        self.entries.iter().find(|(k, _)| k == name).map(|(_, v)| v)
    }

    /// A `uint`-typed entry.
    pub fn uint(&self, name: &str) -> Option<u64> {
        match self.get(name)? {
            Value::Uint(v) => Some(*v),
            _ => None,
        }
    }

    /// A string/blob entry, as bytes.
    pub fn bytes(&self, name: &str) -> Option<&[u8]> {
        match self.get(name)? {
            Value::Bytes(v) => Some(v),
            _ => None,
        }
    }

    /// An array-of-object entry.
    pub fn sections(&self, name: &str) -> Option<&[Section]> {
        match self.get(name)? {
            Value::SectionArray(v) => Some(v),
            _ => None,
        }
    }

    /// The entries, in insertion order.
    pub fn entries(&self) -> &[(String, Value)] {
        &self.entries
    }
}

impl From<u64> for Value {
    fn from(v: u64) -> Value {
        Value::Uint(v)
    }
}
impl From<bool> for Value {
    fn from(v: bool) -> Value {
        Value::Bool(v)
    }
}
impl From<&str> for Value {
    fn from(v: &str) -> Value {
        Value::Bytes(v.as_bytes().to_vec())
    }
}
impl From<Vec<u8>> for Value {
    fn from(v: Vec<u8>) -> Value {
        Value::Bytes(v)
    }
}
impl From<Section> for Value {
    fn from(v: Section) -> Value {
        Value::Section(v)
    }
}
impl From<Vec<u64>> for Value {
    fn from(v: Vec<u64>) -> Value {
        Value::UintArray(v)
    }
}
impl From<Vec<Section>> for Value {
    fn from(v: Vec<Section>) -> Value {
        Value::SectionArray(v)
    }
}

fn append_varint(b: &mut Vec<u8>, v: u64) {
    match v {
        0..=63 => b.push((v << 2) as u8),
        64..=16383 => b.extend_from_slice(&(((v << 2) | 1) as u16).to_le_bytes()),
        16384..=1073741823 => b.extend_from_slice(&(((v << 2) | 2) as u32).to_le_bytes()),
        _ => b.extend_from_slice(&((v << 2) | 3).to_le_bytes()),
    }
}

fn append_string(b: &mut Vec<u8>, s: &[u8]) {
    append_varint(b, s.len() as u64);
    b.extend_from_slice(s);
}

fn append_section(b: &mut Vec<u8>, s: &Section) {
    append_varint(b, s.entries.len() as u64);
    for (k, v) in &s.entries {
        b.push(k.len() as u8);
        b.extend_from_slice(k.as_bytes());
        append_entry(b, v);
    }
}

fn append_entry(b: &mut Vec<u8>, v: &Value) {
    match v {
        Value::Uint(x) => {
            b.push(TYPE_UINT64);
            b.extend_from_slice(&x.to_le_bytes());
        }
        Value::Bool(x) => {
            b.push(TYPE_BOOL);
            b.push(*x as u8);
        }
        Value::Bytes(x) => {
            b.push(TYPE_STRING);
            append_string(b, x);
        }
        Value::Section(x) => {
            b.push(TYPE_OBJECT);
            append_section(b, x);
        }
        Value::UintArray(x) => {
            b.push(TYPE_UINT64 | FLAG_ARRAY);
            append_varint(b, x.len() as u64);
            for e in x {
                b.extend_from_slice(&e.to_le_bytes());
            }
        }
        Value::SectionArray(x) => {
            b.push(TYPE_OBJECT | FLAG_ARRAY);
            append_varint(b, x.len() as u64);
            for e in x {
                append_section(b, e);
            }
        }
        Value::Array(_) => panic!("epee: cannot encode a heterogeneous array"),
    }
}

/// Encodes a root section to the portable-storage binary format.
pub fn marshal(root: &Section) -> Vec<u8> {
    let mut b = Vec::new();
    b.extend_from_slice(&SIGNATURE_A.to_le_bytes());
    b.extend_from_slice(&SIGNATURE_B.to_le_bytes());
    b.push(FORMAT_VER);
    append_section(&mut b, root);
    b
}

struct EpeeReader<'a> {
    b: &'a [u8],
    p: usize,
}

impl<'a> EpeeReader<'a> {
    fn take(&mut self, n: usize) -> Result<&'a [u8]> {
        if self.p + n > self.b.len() {
            return Err(crate::err!(
                "epee: unexpected end of buffer (need {n} at {} of {})",
                self.p,
                self.b.len()
            ));
        }
        let v = &self.b[self.p..self.p + n];
        self.p += n;
        Ok(v)
    }

    fn byte(&mut self) -> Result<u8> {
        Ok(self.take(1)?[0])
    }

    fn varint(&mut self) -> Result<u64> {
        let first = *self.b.get(self.p).ok_or(Error::UnexpectedEof)?;
        Ok(match first & 0x03 {
            0 => self.take(1)?[0] as u64 >> 2,
            1 => u16::from_le_bytes(self.take(2)?.try_into().unwrap()) as u64 >> 2,
            2 => u32::from_le_bytes(self.take(4)?.try_into().unwrap()) as u64 >> 2,
            _ => u64::from_le_bytes(self.take(8)?.try_into().unwrap()) >> 2,
        })
    }

    fn read_section(&mut self) -> Result<Section> {
        let count = self.varint()?;
        let mut s = Section::new();
        for _ in 0..count {
            let nl = self.byte()? as usize;
            let name = String::from_utf8_lossy(self.take(nl)?).into_owned();
            let v = self.read_entry()?;
            s.set(&name, v);
        }
        Ok(s)
    }

    fn read_entry(&mut self) -> Result<Value> {
        let t = self.byte()?;
        if t & FLAG_ARRAY != 0 {
            return self.read_array(t & !FLAG_ARRAY);
        }
        self.read_value(t)
    }

    fn read_array(&mut self, base: u8) -> Result<Value> {
        let count = self.varint()?;
        if base == TYPE_OBJECT {
            let mut out = Vec::new();
            for _ in 0..count {
                out.push(self.read_section()?);
            }
            return Ok(Value::SectionArray(out));
        }
        let mut out = Vec::new();
        for _ in 0..count {
            out.push(self.read_value(base)?);
        }
        if out.iter().all(|v| matches!(v, Value::Uint(_))) {
            return Ok(Value::UintArray(
                out.into_iter()
                    .map(|v| match v {
                        Value::Uint(x) => x,
                        _ => unreachable!(),
                    })
                    .collect(),
            ));
        }
        Ok(Value::Array(out))
    }

    fn read_value(&mut self, t: u8) -> Result<Value> {
        Ok(match t {
            TYPE_UINT64 | TYPE_INT64 => {
                Value::Uint(u64::from_le_bytes(self.take(8)?.try_into().unwrap()))
            }
            TYPE_UINT32 | TYPE_INT32 => {
                Value::Uint(u32::from_le_bytes(self.take(4)?.try_into().unwrap()) as u64)
            }
            TYPE_UINT16 | TYPE_INT16 => {
                Value::Uint(u16::from_le_bytes(self.take(2)?.try_into().unwrap()) as u64)
            }
            TYPE_UINT8 | TYPE_INT8 => Value::Uint(self.byte()? as u64),
            TYPE_BOOL => Value::Bool(self.byte()? != 0),
            // Raw bits; doubles are not used by anything we call.
            TYPE_DOUBLE => Value::Uint(u64::from_le_bytes(self.take(8)?.try_into().unwrap())),
            TYPE_STRING => {
                let n = self.varint()? as usize;
                Value::Bytes(self.take(n)?.to_vec())
            }
            TYPE_OBJECT => Value::Section(self.read_section()?),
            other => return Err(crate::err!("epee: unsupported value type {other}")),
        })
    }
}

/// Decodes a portable-storage binary buffer into its root section.
pub fn unmarshal(data: &[u8]) -> Result<Section> {
    if data.len() < 9 {
        return Err(crate::err!("epee: too short ({} < 9)", data.len()));
    }
    if u32::from_le_bytes(data[0..4].try_into().unwrap()) != SIGNATURE_A
        || u32::from_le_bytes(data[4..8].try_into().unwrap()) != SIGNATURE_B
    {
        return Err(Error::msg("epee: bad signature"));
    }
    if data[8] != FORMAT_VER {
        return Err(crate::err!("epee: unknown format version {}", data[8]));
    }
    let mut r = EpeeReader {
        b: &data[9..],
        p: 0,
    };
    r.read_section()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trip() {
        let mut dist = Section::new();
        dist.set("amount", 0u64)
            .set("global_offsets", vec![1u64, 2, 3]);
        let mut req = Section::new();
        req.set("amounts", vec![dist.clone()])
            .set("height_upper_limit", 0u64)
            .set("use_forced_mix_outs", false)
            .set("coinbase_percents", 0u64)
            .set("blob", vec![0xdeu8, 0xad, 0xbe, 0xef])
            .set("name", "zano");

        let enc = marshal(&req);
        let dec = unmarshal(&enc).unwrap();
        assert_eq!(dec.uint("height_upper_limit"), Some(0));
        assert_eq!(dec.bytes("blob"), Some(&[0xde, 0xad, 0xbe, 0xef][..]));
        assert_eq!(dec.bytes("name"), Some(&b"zano"[..]));
        assert_eq!(dec.get("use_forced_mix_outs"), Some(&Value::Bool(false)));
        let subs = dec.sections("amounts").unwrap();
        assert_eq!(subs.len(), 1);
        assert_eq!(
            subs[0].get("global_offsets"),
            Some(&Value::UintArray(vec![1, 2, 3]))
        );
    }

    #[test]
    fn header_is_validated() {
        assert!(unmarshal(&[]).is_err());
        let mut enc = marshal(&Section::new());
        enc[0] ^= 0xff;
        assert!(unmarshal(&enc).is_err());
        let mut enc = marshal(&Section::new());
        enc[8] = 9;
        assert!(unmarshal(&enc).is_err());
    }

    #[test]
    fn varint_widths() {
        for v in [
            0u64,
            63,
            64,
            16383,
            16384,
            1073741823,
            1073741824,
            u64::MAX >> 2,
        ] {
            let mut b = Vec::new();
            append_varint(&mut b, v);
            let mut r = EpeeReader { b: &b, p: 0 };
            assert_eq!(r.varint().unwrap(), v);
        }
    }

    #[test]
    fn set_replaces_in_place() {
        let mut s = Section::new();
        s.set("a", 1u64).set("b", 2u64).set("a", 3u64);
        assert_eq!(s.entries().len(), 3 - 1);
        assert_eq!(s.uint("a"), Some(3));
    }
}
